lib/puppet/util/windows/service.rb

```diff
@@ -440,23 +440,35 @@ def service_start_type(service_name)
     end
     module_function :service_start_type
 
-    # Change the startup mode of a windows service
+    # Query the configuration of a service using QueryServiceConfigW
+    # to find its current logon account
+    #
+    # @return [String] logon_account account currently set for the service's logon
+    #  in the format "DOMAIN\Account" or ".\Account" if it's a local account
+    def logon_account(service_name)
+      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_QUERY_CONFIG) do |service|
+        query_config(service) do |config|
+          return config[:lpServiceStartName].read_arbitrary_wide_string_up_to(Puppet::Util::Windows::ADSI::User::MAX_USERNAME_LENGTH)
+        end
+      end
+    end
+    module_function :logon_account
+
+    # Set the startup configuration of a windows service
     #
     # @param [String] service_name the name of the service to modify
     # @param [Integer] startup_type a code corresponding to a start type for
     #  windows service, see the "Service start type codes" section in the
     #  Puppet::Util::Windows::Service file for the list of available codes
+    # @param [String] logon_account the account to be used by the service for logon
+    # @param [String] logon_password the provided logon_account's password to be used by the service for logon
     # @param [Bool] delayed whether the service should be started with a delay
-    def set_startup_mode(service_name, startup_type, delayed=false)
-      startup_code = SERVICE_START_TYPES.key(startup_type)
-      if startup_code.nil?
-        raise Puppet::Error.new(_("Unknown start type %{start_type}") % {startup_type: startup_type.to_s})
-      end
+    def set_startup_configuration(service_name, startup_type, logon_account, logon_password, delayed=false)
```

Comment:
I wonder if we should have an options hash passed here, instead of a list of options. It'd be definitely easier to figure out what each parameter does when calling the method (i.e. keep `service_name` and have everything else be part of an options hash).
Buuuut on the other hand it looks like there are many methods with lots of parameters in this area of the code, so you might as well leave it as it is.

Reply:
I could try adding a refactoring maintenance commit, if anyone else agrees on the readability improvement. But I wouldn't want to leave any debts behind and have this be the only method in the file that gets cleaned up.

Reply:
Since the method is changed entirely i would go with the options_hash, in which we can set defaults, and avoid sending `nil` from a lot of calls.